--- a/src/gallium/auxiliary/gallivm/lp_bld_misc.cpp	2024-07-01 14:51:31.493037906 +0200
+++ b/src/gallium/auxiliary/gallivm/lp_bld_misc.cpp	2024-07-01 14:54:51.473698849 +0200
@@ -332,7 +332,7 @@
    llvm::StringMap<bool> features;
    llvm::sys::getHostCPUFeatures(features);
 
-   for (StringMapIterator<bool> f = features.begin();
+   for (llvm::StringMapIterator<bool> f = features.begin();
         f != features.end();
         ++f) {
       MAttrs.push_back(((*f).second ? "+" : "-") + (*f).first().str());
