diff --git a/src/gallium/drivers/zink/zink_context.c b/src/gallium/drivers/zink/zink_context.c
index a2322ec28ca..cb96fbab5bc 100644
--- a/src/gallium/drivers/zink/zink_context.c
+++ b/src/gallium/drivers/zink/zink_context.c
@@ -448,7 +448,7 @@ zink_create_sampler_state(struct pipe_context *pctx,
       sci.flags |= VK_SAMPLER_CREATE_NON_SEAMLESS_CUBE_MAP_BIT_EXT;
    if (state->unnormalized_coords) {
       assert(zink->flags & PIPE_CONTEXT_COMPUTE_ONLY);
-      sci.unnormalizedCoordinates = state->unnormalized_coords;
+      //sci.unnormalizedCoordinates = state->unnormalized_coords;
    }
    sci.magFilter = zink_filter(state->mag_img_filter);
    if (sci.unnormalizedCoordinates)
diff --git a/src/gallium/drivers/zink/zink_screen.c b/src/gallium/drivers/zink/zink_screen.c
index 8337573829a..eaa91ebfe9f 100644
--- a/src/gallium/drivers/zink/zink_screen.c
+++ b/src/gallium/drivers/zink/zink_screen.c
@@ -3323,12 +3323,12 @@ zink_internal_create_screen(const struct pipe_screen_config *config, int64_t dev
 
    screen->instance_info.loader_version = zink_get_loader_version(screen);
    if (config) {
-      driParseConfigFiles(config->options, config->options_info, 0, "zink",
-                          NULL, NULL, NULL, 0, NULL, 0);
-      screen->driconf.dual_color_blend_by_location = driQueryOptionb(config->options, "dual_color_blend_by_location");
+      //driParseConfigFiles(config->options, config->options_info, 0, "zink",
+      //                    NULL, NULL, NULL, 0, NULL, 0);
+      //screen->driconf.dual_color_blend_by_location = driQueryOptionb(config->options, "dual_color_blend_by_location");
       //screen->driconf.inline_uniforms = driQueryOptionb(config->options, "radeonsi_inline_uniforms");
-      screen->driconf.emulate_point_smooth = driQueryOptionb(config->options, "zink_emulate_point_smooth");
-      screen->driconf.zink_shader_object_enable = driQueryOptionb(config->options, "zink_shader_object_enable");
+      //screen->driconf.emulate_point_smooth = driQueryOptionb(config->options, "zink_emulate_point_smooth");
+      //screen->driconf.zink_shader_object_enable = driQueryOptionb(config->options, "zink_shader_object_enable");
    }
 
    if (!zink_create_instance(screen))
diff --git a/src/mesa/main/context.c b/src/mesa/main/context.c
index cbe8b4c0914..7b5b4505a85 100644
--- a/src/mesa/main/context.c
+++ b/src/mesa/main/context.c
@@ -1483,14 +1483,14 @@ _mesa_make_current( struct gl_context *newCtx,
       if (!check_compatible(newCtx, drawBuffer)) {
          _mesa_warning(newCtx,
               "MakeCurrent: incompatible visuals for context and drawbuffer");
-         return GL_FALSE;
+         //return GL_FALSE;
       }
    }
    if (newCtx && readBuffer && newCtx->WinSysReadBuffer != readBuffer) {
       if (!check_compatible(newCtx, readBuffer)) {
          _mesa_warning(newCtx,
               "MakeCurrent: incompatible visuals for context and readbuffer");
-         return GL_FALSE;
+         //return GL_FALSE;
       }
    }
 
